### ISSUE TITLE
User accounts

### ISSUE THREAD

**Issue by temujin9** [flynn/flynn#410]
We need user account management. A full-blown LDAP permission integration would be nice, but SSH keys taking the place of the current `flynn cluster add` stanza (which is currently the shared key to the whole cluster) would be good enough. 

Ideally, the keying should not be cluster-specific, to make migration between clusters easier.

**Comment by kevinob11** [comment]
Am I correct that as of right now there is only one set of credentials to auth to a cluster when you add it via `flynn cluster add`? Is there a way to set up multiple sets of credentials or to reset / change them when someone on my team leaves?

**Comment by temujin9** [comment]
@kevinob11 Currently, you are correct. This is the ticket for fixing that issue, and I believe it is fairly high on their queue. We definitely want it at our company, too.

**Comment by kevinob11** [comment]
Thanks @temujin9 I got worried when I hadn't seen any action here since 2014. Happy to help in any way I can.

**Comment by tjboring** [comment]
Any status on when this one might be done?

**Comment by tjboring** [comment]
Until this is done, anyone have recommendations for dealing with adding new clusters? Currently what I'm doing is this (and it's a pain-in-the-ass):

- Document the cluster add command in an internal wiki
- Put the TLS pin and cluster key in LastPass
- Tell other engineers to get the pin/key from LastPass and run `flynn cluster add`

**Comment by titanous** [comment]
I'm currently primarily working on this, I should have some news soon. That procedure sounds right, though only the key is secret, the TLS pin is not secret.

**Comment by icarlosmendez** [comment]
Where do I acquire this elusive TLS pin? Is it the one that's listed in the documentation on https://flynn.io/docs/cli#adding-clusters? I simply don't see how it would have been created for me locally. I am running the demo in a vagrant box for context.

**Comment by titanous** [comment]
@icarlosmendez It is printed as part of the `flynn cluster add` command, as well as stored in `~/.flynnrc` after running the command.

**Comment by jokeyrhyme** [comment]
Have you considered partnering with something like GitLab for these purposes?

- https://about.gitlab.com/
- https://gitlab.com/gitlab-org/gitlab-ce/tree/master#gitlab

GitLab has:

- user directory and groups
- allows project owners to manage per-user and per-group permissions
- allows users to register and manage their SSH keys
- authenticates access to projects using these SSH keys

What if GitLab was installed in a container, just like the default control-plane's PostgreSQL service?

**Comment by maxlazio** [comment]
Adding to the above comment, GitLab also has LDAP integration which could be leveraged.

**Comment by JobV** [comment]
If you think GitLab makes sense for you or you want to discuss this in a more synchronous way, we're happy to help. You're also free to send me an email at job at gitlab com.

**Comment by titanous** [comment]
I'm currently working on the Flynn authentication/authorization system. We will have a very lightweight built-in user database, but I expect a lot of users will integrate external identity systems. The user policy system is designed to be agnostic to the source of user identity, and the rest of the system is completely agnostic to the source of policy for a given request.

Our first external identity integration will likely be OpenID Connect, which will support user login with a very large variety of external providers, including services like Google Apps. Does GitLab support OIDC? It is rapidly becoming the standard API to do federated identity across applications.

**Comment by temujin9** [comment]
@titanous Is there a ticket I might follow here? Getting this worked into our systems is going to be a priority for me.

**Comment by guestisp** [comment]
Any update on this ? Any ETA ?

**Comment by guestisp** [comment]
Bump

**Comment by guestisp** [comment]
Just to be sure: when this issue will be closed, would be possible to use multiple users each with their own apps? In example, user1 can't show/manage user2's apps ?

**Comment by titanous** [comment]
@guestisp Yes, though we may not have full multitenant security isolation in place when the issue is closed.

**Comment by guestisp** [comment]
Do you have an eta for this?

**Comment by smolinari** [comment]
> Yes, though we may not have full multitenant security isolation in place when the issue is closed.

Could this possibly be a layer added by a Flynn user within the container infrastructure itself? It is sort of a weak example, because it doesn't use containers at all, but I was thinking along the lines of something like Salesforce's Force.com platform. 

Scott

**Comment by temujin9** [comment]
Any ideas on scheduling for this ticket? I've got customer use cases that are blocked by lack of this.

**Comment by koolay** [comment]
+1

**Comment by Cryptophobia** [comment]
Hi @titanous  and @temujin9 ,

User management will be one big factor when evaluating PaaS solutions for us as well. We are looking at a few and only flynn.io doesn't have this feature yet. I don't want to be annoying, but expected ETA on this feature? Can we have at least some approximation --> one month, two months, three months.. from now.

Thanks!

**Comment by temujin9** [comment]
@Cryptophobia I'm not part of the Prime Directive team, so I can't answer that directly

I can say that a good key rotation policy, an existing repository with good access control (S3 works for us), and just a little automation works as a well-secured stopgap, for at least some use-cases.

**Comment by Cryptophobia** [comment]
@temujin9 Thanks for the comment!

So to understand this: you are pushing user's repositories to S3 first and controlling the access to those buckets before deploying them with flynn? Then rotating the keys to make sure they are fresh? 

We want to be able to allow users to push to flynn themselves; do you manage those user-permissions based on the bucket policies/IAM roles? One awesome thing about flynn is the CLI tool and doing the deployments through the git remotes, but I don't see anyway we can give users different level of push (deploy) access within flynn yet.

For you S3 serves as your intermediary access control to flynn?

**Comment by temujin9** [comment]
@Cryptophobia Not quite. We are sharing the Flynn keys via KMS locked S3 plus some interstitial automation, and then users are managing their own applications with the CLI. There is no privilege separation with this solution: access is all or nothing. That works fine in some use cases (basically, if you can trust your developers), but may not be acceptable for yours.

We haven't tried Gitlab at all.

**Comment by Cryptophobia** [comment]
@temujin9  Thanks for the explanation!

**Comment by alizowghi** [comment]
Hi all
Is this feature under development?

**Comment by hadifarnoud** [comment]
I don't think so @alizowghi

**Comment by shinebayar-g** [comment]
ha 4 years, no user authentication. development in a nutshell.

**Comment by Cryptophobia** [comment]
did someone say nutshell? :shipit:

**Comment by titanous** [comment]
Flynn is unmaintained and our infrastructure will shut down on June 1, 2021. See [the README](https://github.com/flynn/flynn#flynn-is-unmaintained) for details.
